### PR TITLE
feat: providers resolve on context change

### PR DESCRIPTION
**Pull request by andreas-karlsson**
#### :heavy_check_mark: Checklist

- [ ] All tests are passing
- [ ] Relevant documentation updated
- [ ] linter/style run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes
- [ ] Tested in a corresponding example app

### CODE REVIEW

File: packages/openfeature-server-provider/src/ConfidenceServerProvider.ts [hunk 1]
```diff
@@ -112,9 +108,7 @@ export class ConfidenceServerProvider implements Provider {
   ): Promise<ResolutionDetails<T>> {
     const [flagName] = flagKey.split('.');
 
-    const configuration = await this.confidence
-      .withContext(this.convertContext(context))
-      .resolve([`flags/${flagName}`]);
+    const configuration = await this.confidence.withContext(convertContext(context)).resolve([`flags/${flagName}`]);
```
Comment (andreas-karlsson) [inline packages/openfeature-server-provider/src/ConfidenceServerProvider.ts:111]:
```suggestion
    const flagResolution = await this.confidence.withContext(convertContext(context)).resolve([`flags/${flagName}`]);
```

Comment (nicklasl) [reply packages/openfeature-server-provider/src/ConfidenceServerProvider.ts]: fixed.

File: packages/openfeature-web-provider/src/ConfidenceWebProvider.ts [hunk 2]
```diff
@@ -14,51 +14,38 @@ import {
...
+      this.events.emit(ProviderEvents.Stale);
+      this.resolve();
+    });
+    await this.resolve();
+  }
+
+  async onClose(): Promise<void> {
+    this.unsubscribe?.();
+    this.flagResolution?.cancel();
+    this.flagResolution = undefined;
```
Comment (andreas-karlsson) [inline packages/openfeature-web-provider/src/ConfidenceWebProvider.ts:48]: Set state not_ready?

File: packages/sdk/src/FlagResolverClient.test.ts [hunk 3]
```diff
@@ -55,7 +55,7 @@ describe('withRequestLogic', () => {
       expect(attempts).toEqual([0, 0, 0, 1000, 2000]);
     });
 
-    it('should abort the previous request', async () => {
+    it.skip('should abort the previous request', async () => {
```
Comment (andreas-karlsson) [inline packages/sdk/src/FlagResolverClient.test.ts:58]: Change test to "PendingFlagReolution can be canceld"

File: packages/openfeature-web-provider/src/ConfidenceWebProvider.test.ts [hunk 4]
```diff
@@ -118,118 +134,144 @@ describe('ConfidenceProvider', () => {
...
+
+      it('should setup a context change subscription', async () => {
+        await instanceUnderTest.initialize({ targetingKey: 'test' });
+        expect(contextChangesMock).toHaveBeenCalledOnce(expect.any(Function));
+      });
     });
 
-    it('should change the provider status with context', async () => {
-      expect(instanceUnderTest.status).toEqual(ProviderStatus.NOT_READY);
+    // it('should change the provider status with context', async () => {
```
Comment (andreas-karlsson) [inline packages/openfeature-web-provider/src/ConfidenceWebProvider.test.ts:162]: Remove

File: packages/openfeature-web-provider/src/ConfidenceWebProvider.ts [hunk 5]
```diff
@@ -67,6 +54,22 @@ export class ConfidenceWebProvider implements Provider {
...
     this.confidence.setContext(convertContext(changes));
+    // following await makes sure that our subscription to context changes has fired (if the context changed)
+    await Promise.resolve();
+    await this.flagResolution;
+  }
+
+  private async resolve(): Promise<void> {
+    try {
+      this.flagResolution?.cancel();
+      this.flagResolution = this.confidence.resolve([]);
```
Comment (andreas-karlsson) [inline packages/openfeature-web-provider/src/ConfidenceWebProvider.ts:65]: Should not overwrite the instance flagResolution until it's ready.

File: packages/openfeature-web-provider/src/ConfidenceWebProvider.ts [hunk 6]
```diff
@@ -16,69 +16,88 @@ import equal from 'fast-deep-equal';
...
+      if (flagResolution) {
+        this.flagResolution = flagResolution;
+        this.events.emit(ProviderEvents.Ready);
+        this.status = ProviderStatus.READY;
+      } else if (this.flagResolution) {
+        this.events.emit(ProviderEvents.Stale);
+        this.status = ProviderStatus.STALE;
+      }
+    });
+    await this.waitForReady().then(isReady => {
```
Comment (nicklasl) [inline packages/openfeature-web-provider/src/ConfidenceWebProvider.ts:48]: nit: use result of `this.confidence.setContext` to skip this await if possible?

File: packages/openfeature-web-provider/src/ConfidenceWebProvider.ts [hunk 7]
```diff
@@ -16,69 +16,88 @@ import equal from 'fast-deep-equal';
...
+      const handler = () => {
+        resolve(true);
+        this.events.removeHandler(ProviderEvents.Ready, handler);
+        clearTimeout(timeoutId);
+      };
+      const timeoutId = setTimeout(() => {
+        resolve(false);
+        this.events.removeHandler(ProviderEvents.Ready, handler);
+      }, this.confidence.config.timeout);
+      this.events.addHandler(ProviderEvents.Ready, handler);
```
Comment (nicklasl) [inline packages/openfeature-web-provider/src/ConfidenceWebProvider.ts:73]: This is kind of nice to use the built in handler.

**Comment by andreas-karlsson** [comment]
Closed in favour of #128

**Pull request closed without merging**